src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/DownloadTimeoutStream.cs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Protocol.Core.v3;
+
+namespace NuGet.Protocol
+{
+    public class DownloadTimeoutStream : Stream
+    {
+        private readonly string _downloadName;
+        private readonly Stream _networkStream;
+        private readonly TimeSpan _timeout;
+
+        public DownloadTimeoutStream(string downloadName, Stream networkStream, TimeSpan timeout)
+        {
+            if (downloadName == null)
+            {
+                throw new ArgumentNullException(nameof(downloadName));
+            }
+
+            if (networkStream == null)
+            {
+                throw new ArgumentNullException(nameof(networkStream));
+            }
+            
+            _downloadName = downloadName;
+            _networkStream = networkStream;
+            _timeout = timeout;
+        }
+
+        public override void Flush()
+        {
+            throw new NotSupportedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            try
+            {
+                return ReadAsync(buffer, offset, count, CancellationToken.None).Result;
+            }
+            catch (AggregateException e)
+            {
+                throw e.InnerException;
+            }
+        }
+
+        public override async Task<int> ReadAsync(
```

Comment:
Our client code does not use `BeginRead`/`EndRead`, so I will thrown a `NotSupportedException`.